packages/ui/src/ui/pages/query-tracker/QueryResultsVisualization/store/reducer.tsx

```diff
@@ -0,0 +1,213 @@
+import type {
+    ChartSettings,
+    ExtendedQueryItem,
+    Field,
+    Visualization,
+    VisualizationId,
+} from '../types';
+import type {ActionD} from '../../../../../ui/types';
+import {QueryResult} from '../preparers/types';
+
+export interface QueryResultsVisualizationState {
+    saved: boolean;
+    query: ExtendedQueryItem | undefined;
+    queryResults: Record<number, QueryResult>;
+    resultIndex: number;
+    visualizations: Visualization[];
+}
+
+function getInitialVisualization(): Visualization {
+    return {
+        id: 'line',
+        placeholders: [
+            {
+                id: 'x',
+                fields: [],
+            },
+            {
+                id: 'y',
+                fields: [],
+            },
+            {
+                id: 'colors',
+                fields: [],
+            },
+        ],
+        chartSettings: {
+            xAxis: {
+                legend: 'on',
+                labels: 'on',
+                title: '',
+                grid: 'on',
+                pixelInterval: '',
+            },
+            yAxis: {
+                labels: 'on',
+                title: '',
+                grid: 'on',
+                pixelInterval: '',
+            },
+        },
+    };
+}
+
+export const initialState: QueryResultsVisualizationState = {
+    saved: true,
+    query: undefined,
+    queryResults: {},
+    resultIndex: 0,
+    visualizations: [getInitialVisualization()],
+};
+
+export type QueryResultVisualizationAction =
+    | ActionD<'set-fields', {placeholderId: string; fields: Field[]}>
+    | ActionD<'remove-field', {placeholderId: string; field: Field}>
+    | ActionD<'set-chart-settings', ChartSettings>
+    | ActionD<'set-visualization', VisualizationId>
+    | ActionD<'set-query', {query: ExtendedQueryItem}>
+    | ActionD<'set-result-index', number>
+    | ActionD<'set-visualization-saved', {saved: boolean}>
+    | ActionD<'set-query-result', {queryResult: QueryResult}>;
+
+export function queryResultsVisualization(
+    state: QueryResultsVisualizationState = initialState,
+    action: QueryResultVisualizationAction,
+): QueryResultsVisualizationState {
+    switch (action.type) {
+        case 'set-visualization-saved': {
+            const {saved} = action.data;
+
+            return {
+                ...state,
+                saved,
+            };
+        }
+        case 'set-query': {
+            const {query} = action.data;
+
+            const queryVisualizations = query.annotations?.ui_chart_config || [];
+
+            const visualizations = [];
+
+            for (let i = 0; i < query.result_count; i++) {
```

Comment:
it would be easier to use map

Reply:
We can do that, but I will need to change all the code around. Could we keep it as is, because now it's already works correctly?

Reply:
ok